File: jackson/src/test/java/org/zalando/guild/api/json/fields/jackson/JsonFieldsModuleTest.java
package org.zalando.guild.api.json.fields.jackson;

import com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility;
import com.fasterxml.jackson.annotation.PropertyAccessor;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.jayway.jsonassert.JsonAssert;
import com.jayway.jsonassert.JsonAsserter;
import org.junit.Before;
import org.junit.Test;
import org.zalando.guild.api.json.fields.java.model.FieldPredicate;

import javax.annotation.Nonnull;
import java.lang.reflect.InvocationHandler;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Supplier;

import static java.lang.reflect.Proxy.newProxyInstance;
import static java.util.concurrent.TimeUnit.NANOSECONDS;
import static org.hamcrest.CoreMatchers.nullValue;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.Is.is;
import static org.zalando.guild.api.json.fields.java.model.FieldPredicates.*;

/**
 * @author Sean Patrick Floyd (sean.floyd@zalando.de)
 * @since 24.09.2015
 */
public class JsonFieldsModuleTest {

    private static final ThreadLocal<FieldPredicate> PREDICATE = new InheritableThreadLocal<>() {
        @Override
        protected FieldPredicate initialValue() {
            return alwaysTrue();
        }
    };

    private ObjectMapper objectMapper;
    private final Outer outer = new Outer();
    private final AtomicInteger runs = new AtomicInteger();

    @Before
    public void setUp() throws Exception {

        objectMapper = new ObjectMapper();
        objectMapper.setVisibility(PropertyAccessor.FIELD, Visibility.ANY);

        final Supplier<FieldPredicate> predicateSupplier = new Supplier<>() {
            @Nonnull
            @Override
            public FieldPredicate get() {
                return PREDICATE.get();
            }
        };

        final ContextProvider contextProvider = ThreadLocalContextProvider.getInstance();
        objectMapper.registerModule(JsonFieldsModule.createJsonFieldsModule(predicateSupplier, contextProvider));
    }

    static class Outer {
        private final Middle foo = new Middle();
        private final String foo2 = "FOO2";
    }

    static class Middle {
        private final Inner bar = new Inner();
        private final int bar2 = 123;

    }

    static class Inner {
        private final String baz = "BAZ";
        private final boolean phleem = true;
    }

    private JsonAsserter asserterFor(final Object obj) {
        try {
            final String json = objectMapper.writeValueAsString(obj);
            System.out.println(json);
            return getJsonAsserter(json);
        } catch (JsonProcessingException e) {
            throw new AssertionError(e);
        }
    }

    /**
     * Create a Proxy around JsonAsserter that rethrows AssertionError but filters out everything else. This is
     * necessary because otherwise JsonAsserter chokes on paths that go deeper than it can follow.
     */
    private JsonAsserter getJsonAsserter(final String json) {
        final JsonAsserter delegate = JsonAssert.with(json);
        final ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
        final Class<?>[] interfaces = {JsonAsserter.class};
        final AtomicReference<JsonAsserter> asserterHolder = new AtomicReference<>();
        final InvocationHandler invocationHandler = new InvocationHandler() {
            @Override
            public Object invoke(final Object proxy, final Method method, final Object[] args) throws Throwable {
                try {
                    method.invoke(delegate, args);
                } catch (InvocationTargetException e) {
                    final Throwable cause = e.getCause();
                    if (cause instanceof AssertionError) {
                        throw cause;
                    }
                }

                // all methods are fluid, so always return the proxy instance.
                return asserterHolder.get();
            }
        };

        final JsonAsserter jsonAsserter = (JsonAsserter) newProxyInstance(classLoader, interfaces, invocationHandler);
        asserterHolder.set(jsonAsserter);
        return jsonAsserter;
    }

    @Test
    public void singleThreadEnvironment() {
        task(1000).run();
    }

    @Test
    public void multiThreadedEnvironment() throws InterruptedException {
        final int threads = Runtime.getRuntime().availableProcessors() * 2;
        final int runsPerThread = 100;
        final int totalRuns = threads * runsPerThread;

        runs.set(0);

        final ExecutorService executorService = Executors.newFixedThreadPool(threads);
        final Runnable task = task(runsPerThread);
        for (int i = 0; i < threads; i++) {
            executorService.submit(task);
        }

        executorService.shutdown();
        assertThat(executorService.awaitTermination(Long.MAX_VALUE, NANOSECONDS), is(true));
        assertThat(runs.get(), is(totalRuns));

    }

    private CountingRunner task(final int runsPerThread) {
        return new CountingRunner(runsPerThread);
    }

    private class CountingRunner implements Runnable {

        private final int runsPerThread;

        public CountingRunner(final int runsPerThread) {
            this.runsPerThread = runsPerThread;
        }

        @Override
        public void run() {
            for (int i = 0; i < runsPerThread; i++) {
                doRun();
            }
        }

        private void doRun() {

            allFieldsEnabled();
            noFieldsEnabled();
            simpleMatch();
            complicatedMatch();

            runs.incrementAndGet();
        }

        private void complicatedMatch() {
            PREDICATE.set(and(not(matchIndex(1, "bar")),
                    not(matchIndex(0, "foo2"))));
            asserterFor(outer).assertThat("$.foo2", is(nullValue()))
                    .assertThat("$.foo.bar", is(nullValue()))
                    .assertThat("$.foo.bar2", is(123));
        }

        private void simpleMatch() {
            PREDICATE.set(matchIndex(1, "bar"));
            asserterFor(outer).assertThat("$.foo2", is("FOO2"))
                    .assertThat("$.foo.bar.baz", is("BAZ"))
                    .assertThat("$.foo.bar2", is(nullValue()))
                    .assertThat("$.foo.bar.phleem", is(true));
        }

        private void noFieldsEnabled() {
            PREDICATE.set(alwaysFalse());
            asserterFor(outer).assertThat("$.foo", is(nullValue()))
                    .assertThat("$.foo2", is(nullValue()));
        }

        private void allFieldsEnabled() {
            PREDICATE.set(alwaysTrue());
            asserterFor(outer).assertThat("$.foo2", is("FOO2"))
                    .assertThat("$.foo.bar.baz", is("BAZ"))
                    .assertThat("$.foo.bar2", is(123))
                    .assertThat("$.foo.bar.phleem", is(true));
        }

    }
}

File: jackson/src/test/java/org/zalando/guild/api/json/fields/jackson/JsonFieldsModuleTest.kt
package org.zalando.guild.api.json.fields.jackson

import com.fasterxml.jackson.annotation.JsonAutoDetect
import com.fasterxml.jackson.annotation.PropertyAccessor
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.zalando.guild.api.json.fields.java.model.FieldPredicate
import org.zalando.guild.api.json.fields.java.model.FieldPredicates.alwaysFalse
import org.zalando.guild.api.json.fields.java.model.FieldPredicates.alwaysTrue
import org.zalando.guild.api.json.fields.java.model.FieldPredicates.and
import org.zalando.guild.api.json.fields.java.model.FieldPredicates.matchIndex
import org.zalando.guild.api.json.fields.java.model.FieldPredicates.not
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Supplier

class JsonFieldsModuleTest {
    private var objectMapper: ObjectMapper? = null
    private val outer = Outer()
    private val runs = AtomicInteger()

    @BeforeEach
    fun setUp() {
        objectMapper = ObjectMapper()
        objectMapper!!.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY)
        val predicateSupplier = Supplier { PREDICATE.get() }
        objectMapper!!.registerModule(JsonFieldsModule.createJsonFieldsModule(predicateSupplier))
    }

    internal class Outer {
        private val foo = Middle()
        private val foo2 = "FOO2"
    }

    internal class Middle {
        private val bar = Inner()
        private val bar2 = 123
    }

    internal class Inner {
        private val baz = "BAZ"
        private val phleem = true
    }

    private fun serialize(obj: Any): JsonNode {
        val json = objectMapper!!.writeValueAsString(obj)
        return objectMapper!!.readTree(json)
    }

    @Test
    fun singleThreadEnvironment() {
        task(1000).run()
    }

    @Test
    fun multiThreadedEnvironment() {
        val threads = Runtime.getRuntime().availableProcessors() * 2
        val runsPerThread = 100
        val totalRuns = threads * runsPerThread
        runs.set(0)
        val executorService = Executors.newFixedThreadPool(threads)
        val task: Runnable = task(runsPerThread)
        for (i in 0 until threads) {
            executorService.submit(task)
        }
        executorService.shutdown()
        assertTrue(executorService.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS))
        assertEquals(totalRuns, runs.get())
    }

    private fun task(runsPerThread: Int): CountingRunner {
        return CountingRunner(runsPerThread)
    }

    private inner class CountingRunner(private val runsPerThread: Int) : Runnable {
        override fun run() {
            for (i in 0 until runsPerThread) {
                doRun()
            }
        }

        private fun doRun() {
            allFieldsEnabled()
            noFieldsEnabled()
            simpleMatch()
            complicatedMatch()
            runs.incrementAndGet()
        }

        private fun complicatedMatch() {
            PREDICATE.set(
                and(
                    not(matchIndex(1, "bar")),
                    not(matchIndex(0, "foo2"))
                )
            )
            val json = serialize(outer)
            assertFalse(json.has("foo2"))
            // foo is also excluded: NOT(matchIndex(1, "bar")) fails for ["foo"]
            // because matchIndex vacuously passes when hierarchy is too short
            assertFalse(json.has("foo"))
        }

        private fun simpleMatch() {
            PREDICATE.set(matchIndex(1, "bar"))
            val json = serialize(outer)
            assertEquals("FOO2", json["foo2"].textValue())
            assertEquals("BAZ", json["foo"]["bar"]["baz"].textValue())
            assertFalse(json["foo"].has("bar2"))
            assertEquals(true, json["foo"]["bar"]["phleem"].booleanValue())
        }

        private fun noFieldsEnabled() {
            PREDICATE.set(alwaysFalse())
            val json = serialize(outer)
            assertFalse(json.has("foo"))
            assertFalse(json.has("foo2"))
        }

        private fun allFieldsEnabled() {
            PREDICATE.set(alwaysTrue())
            val json = serialize(outer)
            assertEquals("FOO2", json["foo2"].textValue())
            assertEquals("BAZ", json["foo"]["bar"]["baz"].textValue())
            assertEquals(123, json["foo"]["bar2"].intValue())
            assertEquals(true, json["foo"]["bar"]["phleem"].booleanValue())
        }
    }

    companion object {
        private val PREDICATE: ThreadLocal<FieldPredicate> = object : InheritableThreadLocal<FieldPredicate>() {
            override fun initialValue(): FieldPredicate = alwaysTrue()
        }
    }
}
